java/AxivityReader.java

```diff
@@ -348,5 +360,25 @@ private static LocalDateTime cwaHeaderLoggingStartTime(ByteBuffer buf) {
     }
 
 
+    private static void setTimeZone(String timeZone) {
+        zoneId = ZoneId.of(timeZone);
+        rules = zoneId.getRules();
+    }
+
+
+    private static void setSessionStart(LocalDateTime ldt) {
+        sessionStart = ldt.atZone(zoneId);
+        sessionStartDST = rules.isDaylightSavings(sessionStart.toInstant());
+    }
+
+
+    private static ZonedDateTime zonedWithDSTCorrection(LocalDateTime ldt) {
+        ZonedDateTime zdt = ldt.atZone(zoneId);
+        boolean isDST = rules.isDaylightSavings(zdt.toInstant());
+        if (isDST != sessionStartDST) {  // DST crossover happened
+            zdt = isDST ? ldt.plusHours(1).atZone(zoneId) : zdt.plusHours(-1);
+        }
+        return zdt;
+    }
```

Comment:
At last, the DST crossover correction code. It takes a local time (no timezone), zones it to the user-specified timezone (e.g. Europe/London), checks to see if DST has changed (wrt. session start), and if so correct accordingly.

It is perhaps worth noting why forward and backward shifts are done differently. When stepping _into_ DST `ZonedDateTime` is smart enough to automatically push forward local times that lie in the inexistent hour, for example `LocalDateTime.parse("2019-03-31`**`T01:05`**`").atZone(ZoneId.of("Europe/London")` automagically produces `2019-03-31T`**`02:05+01:00`**`[Europe/London]` because push forward happened at 01:00 and so 01:05 doesn't really exist. This is great functionality from `ZonedDateTime`, but the same will not happen with hours that may well exist, e.g. `LocalDateTime.parse("2019-03-31T`**`02:05`**`").atZone(ZoneId.of("Europe/London")`  also produces the same timestamp `2019-03-31T`**`02:05+01:00`**`[Europe/London]`. To solve this we just manually push forward the local hour _before_ zoning.

When stepping _out of_ DST, a different situation happens. For example `LocalDateTime.parse("2019-10-27`**`T01:05`**`").atZone(ZoneId.of("Europe/London")` is really an ambiguous time (is this 01:05 _before_ the push back at 02:00 or _after_?), but the output of such line is `2019-10-27T`**`01:05+01:00`**`[Europe/London]`, i.e. `ZonedDateTime` prefers to keep the DST status before the pushback. So using `LocalDateTime` there's no way to obtain something like `2019-10-27T`**`01:05+00:00`**`[Europe/London]` i.e. 01:05 _after_ push back. To solve this we manually push back  _after_ the zoning, so to obtain the above time we first do `LocalDateTime.parse("2019-10-27`**`T02:05`**`").atZone(ZoneId.of("Europe/London")` which first produces `2019-10-27T`**`02:05+00:00`**`[Europe/London]`, then `.plusHours(-1)` to produce `2019-10-27T`**`01:05+00:00`**`[Europe/London]`.

Reply:
Can you edit the text in this comment to move from the somewhat confusing (to me at least!) 'push forward'/'push back' terminology to 'spring forward'/'fall back' (more commonly used to describe daylight saving changes)?

I love this comment, and we need a way to integrate it into the wider documentation for this tool (I'll look into this) - this is really helpful.

Reply:
Thanks Aiden, I improved the comments as per the suggestions. I hope it is clearer now.